experimental/packages/shim-opencensus/src/metric-transform.ts

```diff
@@ -0,0 +1,209 @@
+  type: oc.MetricDescriptorType
+): MappedType | null {
+  switch (type) {
+    case oc.MetricDescriptorType.GAUGE_INT64:
+      return {
+        type: InstrumentType.OBSERVABLE_GAUGE,
+        valueType: ValueType.INT,
+        dataPointType: DataPointType.GAUGE,
+      };
+    case oc.MetricDescriptorType.GAUGE_DOUBLE:
+      return {
+        type: InstrumentType.OBSERVABLE_GAUGE,
+        valueType: ValueType.DOUBLE,
+        dataPointType: DataPointType.GAUGE,
+      };
+
+    case oc.MetricDescriptorType.CUMULATIVE_INT64:
+      return {
+        type: InstrumentType.COUNTER,
+        valueType: ValueType.INT,
+        dataPointType: DataPointType.SUM,
+      };
+    case oc.MetricDescriptorType.CUMULATIVE_DOUBLE:
+      return {
+        type: InstrumentType.COUNTER,
+        valueType: ValueType.DOUBLE,
+        dataPointType: DataPointType.SUM,
+      };
+
+    case oc.MetricDescriptorType.CUMULATIVE_DISTRIBUTION:
+      return {
+        type: InstrumentType.HISTOGRAM,
+        valueType: ValueType.DOUBLE,
+        dataPointType: DataPointType.HISTOGRAM,
+      };
+
+    case oc.MetricDescriptorType.SUMMARY:
+    case oc.MetricDescriptorType.GAUGE_DISTRIBUTION:
+    case oc.MetricDescriptorType.UNSPECIFIED:
+      diag.warn(
+        'Got unsupported metric MetricDescriptorType from OpenCensus: %s',
+        type
+      );
+      return null;
+  }
+}
+
+function gauge(
+  metric: oc.Metric,
+  dataPointType: DataPointType.GAUGE,
+  baseMetric: BaseMetric
+): GaugeMetricData {
+  return {
+    ...baseMetric,
+    dataPoints: dataPoints(metric, value => value as number),
+    dataPointType,
+  };
+}
+
+function sum(
+  metric: oc.Metric,
+  dataPointType: DataPointType.SUM,
+  baseMetric: BaseMetric
+): SumMetricData {
+  return {
+    ...baseMetric,
+    dataPoints: dataPoints(metric, value => value as number),
+    isMonotonic: true,
+    dataPointType,
+  };
+}
+
+function histogram(
+  metric: oc.Metric,
+  dataPointType: DataPointType.HISTOGRAM,
+  baseMetric: BaseMetric
+): HistogramMetricData {
+  return {
+    ...baseMetric,
+    dataPoints: dataPoints(metric, value => {
+      const {
+        bucketOptions: {
+          explicit: { bounds },
+        },
+        buckets,
+        count,
+        sum: distSum,
+      } = value as oc.DistributionValue;
+
+      return {
+        buckets: {
+          boundaries: bounds,
+          counts: buckets.map(bucket => bucket.count),
+        },
+        count,
+        sum: distSum,
+      };
+    }),
+    dataPointType,
+  };
+}
+
+function dataPoints<T>(
+  metric: oc.Metric,
+  valueMapper: (value: oc.TimeSeriesPoint['value']) => T
+): DataPoint<T>[] {
+  return metric.timeseries.flatMap(ts => {
+    const attributes = zipOcLabels(metric.descriptor.labelKeys, ts.labelValues);
+
+    // use zeroed hrTime if it is undefined, which probably shouldn't happen
+    const startTime = ocTimestampToHrTime(ts.startTimestamp) ?? ZEROED_HRTIME;
+
+    return ts.points.map(
```

Comment:
Does this mean that in one single read of the metric, the exported data might contain duplicated `metricDescriptor, attributeSet` pairs? For example, when compositing with Prometheus exporter, would the following output be possible?

```
my_counter{attr1="value1"} 1
my_counter{attr1="value1"} 2
```

Reply:
I think that in practice, OpenCensus will behave like OTel and only produce one point per unique attribute set. I can double check.

Does that answer your question or did you mean more in terms of conflicts across different MetricProducers in the composite set of metrics?

Reply:
Yeah, if OpenCensus only produces one point per unique attribute set, it solves my question. Thanks for clarifying. 

Would you mind adding a comment here referencing what the reality is, despite the data structure allowing duplicated attribute sets?

Reply:
Sorry if my comment above is misleading. I meant that it would be great if this could be documented in the code.

Reply:
My bad, fixed. Sorry for the confusion.